packages/build/src/plugins_core/deploy/buildbot_client.js

```diff
@@ -0,0 +1,86 @@
+const net = require('net')
+const { promisify } = require('util')
+
+const pEvent = require('p-event')
+
+const { addErrorInfo } = require('../../error/info')
+
+const startBuildbotClient = async function(buildbotServerSocket) {
+  if (buildbotServerSocket === undefined) {
+    return
+  }
+  try {
+    const buildbotClient = net.createConnection(buildbotServerSocket)
+    await pEvent(buildbotClient, 'connect')
+    return buildbotClient
+  } catch (error) {
+    addErrorInfo(error, { type: 'buildbotClientConnection' })
+    throw error
+  }
+}
+
+const closeBuildbotClient = async function(buildbotClient) {
+  if (buildbotClient === undefined) {
+    return
+  }
+  try {
+    await promisify(buildbotClient.end.bind(buildbotClient))()
+  } catch (error) {
+    addErrorInfo(error, { type: 'buildbotClientConnection' })
+    throw error
+  }
+}
+
+const writePayload = async function(buildbotClient, payload) {
+  if (buildbotClient === undefined) {
+    return
+  }
+  try {
+    await promisify(buildbotClient.write.bind(buildbotClient))(JSON.stringify(payload))
+  } catch (e) {
+    const error = new Error(`Error writing TCP payload to buildbot: ${e.message}`)
+    addErrorInfo(error, { type: 'buildbotClient ' })
+    throw error
+  }
+}
+
+const getNextParsedResponsePromise = async function(buildbotClient) {
+  if (buildbotClient === undefined) {
+    return
+  }
+  const data = await pEvent(buildbotClient, 'data')
+  try {
+    return JSON.parse(data)
+  } catch (e) {
+    const error = new Error(`Invalid TCP payload received from the buildbot: ${e.message}\n${data}`)
+    addErrorInfo(error, { type: 'buildbotClient' })
+    throw error
+  }
+}
+
+const deploySiteWithBuildbotClient = async function(client) {
+  const payload = { action: 'deploySite' }
+  try {
+    const nextResponsePromise = getNextParsedResponsePromise(client)
+    await writePayload(client, payload)
+    const response = await nextResponsePromise
+
+    if (!response.succeeded) {
+      throw new Error(`Deploy did not succeed: ${response.values.error}`)
+    }
+  } catch (error) {
+    addErrorInfo(error, {
```

Comment:
Should we merge the two error types into one, or do you think separating them is better?

Reply:
Hmm - good question. I think it makes sense to distinguish the two, because one is a failure to establish the connection at all, and the other is a failure within the buildbot that is simply communicated over the connection. I would be fine combining them, though, since we're only using this for one purpose right now.

Reply:
Actually, if we add a TCP timeout, there won't be a simple way to distinguish between errors in the connection and errors caused by an improper response. Given that, I think it makes sense to combine the two.